### PR TITLE
Use default connection id for KubernetesPodOperator

### LINKED ISSUE #30324: KPO deferrable needs kubernetes_conn_id while non deferrable does not
### Apache Airflow version

2.5.2

### What happened

Not sure if this is a feature not a bug, but I can use KubernetesPodOperator fine without setting a kubernetes_conn_id.
For example:
```
start = KubernetesPodOperator(
        namespace="mynamespace",
        cluster_context="mycontext",
        security_context={ 'runAsUser': 1000 },
        name="hello",
        image="busybox",
        image_pull_secrets=[k8s.V1LocalObjectReference('prodregistry')],
        cmds=["sh", "-cx"],
        arguments=["echo Start"],
        task_id="Start",
        in_cluster=False,
        is_delete_operator_pod=True,
        config_file="/home/airflow/.kube/config",
    )
```

But if I add deferrable=True to this it won't work. It seems to require an explicit kubernetes_conn_id (which we don't configure).

Is not possible to the deferrable version to work as the non deferrable one?

### What you think should happen instead

I hoped that kpo deferrable would work the same as non deferrable.

### How to reproduce

Use KPO with deferrable=True but no kubernetes_conn_id setting

### Operating System

Debian 11

### Versions of Apache Airflow Providers

_No response_

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by lwyszomi**
Fixes: #30324 

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by uranusjr — APPROVED** [review]

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py
Comment (uranusjr) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]: Actually, it’s probably better to use `KubernetesHook.default_conn_name` as the default instead.

Comment (lwyszomi) [reply airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]: @uranusjr I changed and now we using `KubernetesHook.default_conn_name`, thanks

**Comment by jedcunningham** [comment]
@dstandish, do you remember why we didn't do this initially?

**Comment by potiuk** [comment]
> @dstandish, do you remember why we didn't do this initially?

My question exactly.

**Comment by dstandish** [comment]
> > @dstandish, do you remember why we didn't do this initially?
> 
> My question exactly.

Backward compatibility. Previously, KPO did not use airflow conn.  If we now use default, this will break things for users.  We can do but must be major release.

**Comment by potiuk** [comment]
> Backward compatibility. Previously, KPO did not use airflow conn. If we now use default, this will break things for users. We can do but must be major release.

So it was using the bult-in "local" kubernetes (basically whatever kubectl sees ?) if None is used?  

Maybe we can check for existence of the connection and fallback to default behaviour when it is not available? What's the likelihood someone will have default k8s connection set and would not want to use it ?  (still likely requring major version bump and technically breaking but with far less radius blast)

**Comment by dstandish** [comment]
> > Backward compatibility. Previously, KPO did not use airflow conn. If we now use default, this will break things for users. We can do but must be major release.
> 
> So it was using the bult-in "local" kubernetes (basically whatever kubectl sees ?) if None is used?
> 
> Maybe we can check for existence of the connection and fallback to default behaviour when it is not available? What's the likelihood someone will have default k8s connection set and would not want to use it ? (still likely requring major version bump and technically breaking but with far less radius blast)

i think quite likely: https://github.com/apache/airflow/blob/6d09fc7b88623919b01cfbdee5566598cefba021/airflow/utils/db.py#L364-L370

**Comment by dstandish** [comment]
> So it was using the bult-in "local" kubernetes (basically whatever kubectl sees ?) if None is used?

it was using the settings from core, i.e. `get_kube_client`

**Comment by potiuk** [comment]
> i think quite likely:
> 
> https://github.com/apache/airflow/blob/6d09fc7b88623919b01cfbdee5566598cefba021/airflow/utils/db.py#L364-L370

Let me modify it **slightly**. Why don't we fall-back to `get_kube_client` when either the connection is not defined or when it is not **configured** (i.e no extras defined). I believe the only used connection fields are "extras" and it would require an effort to edit the connection to add any extra there - the default one will have no extras at all and we can easily detect it. A bit long shot, but maybe it is the most sensible thing to do? (just wild thought)

**Comment by dstandish** [comment]
Hard no with falling back to get_kube_client -- no reason to bring that back. The kube hook with works with no conn Id and it's fully back compat with get kube client

Now should we ever disregard the default connection? Not sure about that

File: kubernetes_tests/test_kubernetes_pod_operator.py
Comment (uranusjr) [inline kubernetes_tests/test_kubernetes_pod_operator.py]: Doesn’t feel like a good idea to do this. We should instead modify the tests (maybe use mocks) to actually test the conn id.

**Review by dstandish — DISMISSED** [review]
just want to block merge (for now) to allow some time for debate to play out re whether this is good idea. ultimately i do *not* veto this just want to take a moment

File: kubernetes_tests/test_kubernetes_pod_operator.py
Comment (uranusjr) [inline kubernetes_tests/test_kubernetes_pod_operator.py]:
```suggestion
    @mock.patch(f"{HOOK_CLASS}.get_connection", return_value=Connection(conn_id="kubernetes_default"))
    def test_do_xcom_push_defaults_false(self, get_connection_mock):
```

I think you can do this instead.

Since this is repeated so may times, it may be a good idea to make this a fixture instead:

```python
@pytest.fixture()
def mock_get_connection():
    with mock.patch(f"{HOOK_CLASS}.get_connection", return_value=Connection(conn_id="kubernetes_default")):
        yield

@pytest.mark.usefixtures("mock_get_connection")
def test_do_xcom_push_defaults_false(self):
    ...
```

**Comment by dimberman** [comment]
Hey @dstandish @potiuk can we restart this conversation? I wouldn't want this PR to fall into the abyss just because we're unsure. 

FWIW I think it's fine for us to make this breaking change and then do a major update. to me this is where the magic of providers exists. If we want to release a version before that contains a deprecation warning/write a warning somewhere just in case I'm not against that, but I'd love to bring the KPO more in line with the rest of the operators.

**Comment by eladkal** [comment]
>  If we want to release a version before that contains a deprecation warning/write a warning somewhere just in case I'm not against that, but I'd love to bring the KPO more in line with the rest of the operators.

That is preferred

**Comment by dstandish** [comment]
I'm fine with using `kubernetes_default` by default, just should be in a major.  And i don't have an opinion re whether we need to introduce warning first so I defer to y'all on that.

**Comment by eladkal** [comment]
So if decided to go with major release directly (which is OK)
please update provider.yaml to version 6.0.0 and add entry in the provider change log

**Comment by potiuk** [comment]
Needs conflict resplution now, I am afraid

**Comment by eladkal** [comment]
@lwyszomi can you rebase and resolve conflicts?

**Review by eladkal — APPROVED** [review]

File: airflow/providers/cncf/kubernetes/utils/pod_manager.py
Comment (jedcunningham) [inline airflow/providers/cncf/kubernetes/utils/pod_manager.py]:
```suggestion
        kube_client: client.CoreV1Api,
```

Shouldn't this be required now?

File: airflow/providers/cncf/kubernetes/CHANGELOG.rst
Comment (eladkal) [inline airflow/providers/cncf/kubernetes/CHANGELOG.rst]:
```suggestion
Breaking changes
~~~~~~~~~~~~~~~~

Use kubernetes_default connection by default in the KubernetesPodOperator.
```

**Review by hussein-awala — APPROVED** [review]

**Review by eladkal — APPROVED** [review]
cc @jedcunningham

File: kubernetes_tests/test_kubernetes_pod_operator.py
Comment (potiuk) [inline kubernetes_tests/test_kubernetes_pod_operator.py]: Usefixtures makes little sense when put as method decorator (more typing, less explicit, only saves "unused parameter" in IDE. But it makes whole lot of sense when put on a class - can we please remove it it from the methods and put it once on TestKubernetesPodOperatorSystem instead?

**Review by potiuk — APPROVED** [review]

**Comment by iJanki-gr** [comment]
Has this change actually got into 2.6.0 release? I had to change KPO conf to include `kubernetes_conn_id=None,` or I would get The conn_id `kubernetes_default` isn't defined

**Comment by eladkal** [comment]
> Has this change actually got into 2.6.0 release? 

This is not a core PR It's not released with apache-airlfow it's released with Kubernetes provider: `apache-airflow-providers-cncf-kubernetes=6.0.0`

**Comment by iJanki-gr** [comment]
Ok but apparently it was bundled with 2.6.0 https://raw.githubusercontent.com/apache/airflow/constraints-2.6.0/constraints-3.10.txt
And in the docker image as I'm using that.

**Comment by potiuk** [comment]
> Ok but apparently it was bundled with 2.6.0 https://raw.githubusercontent.com/apache/airflow/constraints-2.6.0/constraints-3.10.txt And in the docker image as I'm using that.

From you message it's not clear if you have problem with or not. and what you are asking for. But no matter what - if you think you have a problem and needs older version of the provider, you can alwas build your own custom image where the provider is [downgraded](https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html#installing-upgrading-downgrading-providers-separately-from-airflow-core). The image we publish is the reference image and you are free to [extend it](https://airflow.apache.org/docs/docker-stack/build.html#extending-the-image) as you see fit.

What's your ask really @iJanki-gr ? Since you've already checked the constraints? Is your question answered or do you expect something else here?

**Comment by eladkal** [comment]
This is a breaking change that replaces the default value of the parameter. hence why it was released in a major version (6.0.0) so yet if you don't have `kubernetes_default` define it will break your dags. You need to go trough the change log when you are updating major versions of providers.

**Comment by iJanki-gr** [comment]
Yes that's ok, but in the release notes of airflow 2.6.0 there was no mention of cncf kubernetes provider upgrade.

**Comment by iJanki-gr** [comment]
I fixed my issue, next time i'll check the providers release notes too. Thanks

**Comment by potiuk** [comment]
> Yes that's ok, but in the release notes of airflow 2.6.0 there was no mention of cncf kubernetes provider upgrade.

Because the image bundles airflow 2.6.0 AND providers. The image is a reference one that contains airflow and latest provders, but you are in control on which providers you want to keep or use.  As @eladkal mentioned, if you chose to use the image, you need to check the release notes of the providers.

**Comment by iJanki-gr** [comment]
I will thank you and sorry for the bother.

**Comment by potiuk** [comment]
The kubernetes provider as of 6.0.0 releases uses kubernetes connection to be defined https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/index.html#breaking-changes and you should switch to it. That was a breaking change in the provider. Have you checked the release notes ?

**Comment by potiuk** [comment]
The whole discussion above explaines  all about this in case you have not read it @bmoon4

**Comment by bmoon4** [comment]
i added `kubernetes_conn_id=None` explicitly in `KubernetesPodOperator` parameter and seems working :)

**Comment by dstandish** [comment]
I offer a solution in https://github.com/apache/airflow/pull/31187

If kubernetes_default doesn't exist, ignore it.  If other conn id doesn't exist, fail.

**Pull request merged**
